sentSpawn functionality

deltamolfar opened the pull request:
(Re-PR because I accidentally deleted previous fork)
Added ``:GetBool()``, and review changes.


- Finally a way for E2 to spawn sents.
- For QoL, limited to wire/native gmod sents + entity tab entities, if `wire_expression2_propcore_whitelist <= 0`)
- Thirdparty addons support (ability to add/remove ents to be able to be spawned within whitelist + a hook with same logic as `Expression2_CanEffect`
- Entity registration is straightforward, and the sentSpawn does all the E2 to Lua castings itself (No need to put every check in _preFactory'ies).
- Ability to spawn **ALL** wire entities (if I didn't forget any).
- Few helper functions to sentSpawn.

While most of registration code/logic is from starfall `prop_sent.lua`, I've found and fixed ~10 sents + added ~6 that weren't originally there.


Current minor bugs:
- gmod_wire_dynamic_button doesn't spawn correctly (commented out).
- gmod_cameraprop and gmod_emitter spawn with no issue, but don't work as expected (commented out).
- Undo for some reason doesn't return `E2 sent (classname)`, but `e2_spawned_sent` when user cancels sent.

thegrb93 commented on lua/entities/gmod_wire_expression2/core/custom/prop.lua:
```diff
@@ -194,8 +194,402 @@ local function boneVerify(self, bone)
+		if e2TypeID == TYPE_TABLE then -- Png parameters support
+			if #e2Value ~= 2 then return nil end
+
+			if TypeID(e2Value[1]) ~= TYPE_STRING then return nil end
+			if TypeID(e2Value[2]) ~= TYPE_STRING then return nil end
+
+			return Material(e2Value[1], e2Value[2])
+		end
+
+		return nil
+	end,
+	[TYPE_MATRIX] = function(e2Value) -- TYPE_MATRIX from 'matrix4'
+		if TypeID(e2Value) ~= TYPE_TABLE then return nil end
+
+		if #e2Value == 16 then
+			for i = 1, 16 do
+				if not isnumber(e2Value[i]) then return nil end
+			end
+
+			return Matrix({e2Value[1], e2Value[2], e2Value[3], e2Value[4]}, {e2Value[5], e2Value[6], e2Value[7], e2Value[8]}, {e2Value[9], e2Value[10], e2Value[11], e2Value[12]}, {e2Value[13], e2Value[14], e2Value[15], e2Value[16]})
+		end
+
+		return nil
+	end,
+	[TYPE_COLOR] = function(e2Value) -- TYPE_COLOR from 'vector' or 'vector4' or 'itable' or 'table'
+		local e2TypeID = TypeID(e2Value)
+		if e2TypeID == TYPE_VECTOR then
+			return Color(e2Value[1], e2Value[2], e2Value[3])
+		elseif e2TypeID == TYPE_TABLE then -- vector4 support + direct table support
+			if isnumber(e2Value[1]) and isnumber(e2Value[2]) and isnumber(e2Value[3]) and isnumber(e2Value[4]) then
+				return Color(e2Value[1], e2Value[2], e2Value[3], e2Value[4])
+			elseif e2Value.r and e2Value.g and e2Value.b then
+				if e2Value.a then return Color(e2Value.r, e2Value.g, e2Value.b, e2Value.a) end
+				return Color(e2Value.r, e2Value.g, e2Value.b)
+			end
+		end
+
+		return nil
+	end,
+}
```
This should be e2 core code rather than local to prop lib.

deltamolfar replied:
Yeah, I wanted to make it E2Lib code, but wasn't sure, and never got reply in Discord. Will move it

thegrb93 commented:
Looks fine, might do a diff against starfall's to sanity check.

thegrb93 commented:
I think sents_default_params.lua needs to be put somewhere outside of lua/entities. putting lua files there registers the filename as an entity class.

thegrb93 reviewed (changes requested):
I think sents_default_params.lua needs to be put somewhere outside of lua/entities. putting lua files there registers the filename as an entity class.

deltamolfar commented:
- Alright, I've moved it and renamed sents_default_params.lua out of entities folder, to lua/wire/server/sents_registry.lua.
- Changed a whitelist/registry logic a bit (reinforced the registry reusability by thirdparty addons, by separating registering wire elements, and blacklisting).
- Few QoL fixes.

thegrb93 commented on lua/wire/server/sents_registry.lua:
```diff
@@ -33,43 +38,56 @@
 
-	local whitelist = list.GetForEdit("wire_spawnable_ents_whitelist")
+	local whitelist = list.GetForEdit("wire_spawnable_ents_registry")
 	table.remove(whitelist, class)
```
Shouldn't this be `whitelist[class]=nil`? I don't think table.remove works on string indexed tables.

deltamolfar replied:
Yeah you're right, my bad
